C# 12: Inline array support.

michaelnebel opened the pull request:
In this PR we make support for *inline arrays*.
The code sample below contains the declaration of the *inline array* type `MyInlineArray`, which can contain 10 elements.
Inline arrays are declared using structs decorated with the `InlineArray` attribute.
```csharp
[System.Runtime.CompilerServices.InlineArray(10)]
public struct MyInlineArray
{
    private int myInlineArrayElements;
}

public class Program
{
    public static void Main(string[] args)
    {
        var arr = new MyInlineArray();
        for (int i = 0; i < 10; i++)
        {
            arr[i] = i + 1;
        }
    }
}
```
Microsoft has a short description of the new language feature [here](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#inline-arrays) and the more detailed feature specification can be seen [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-12.0/inline-arrays).

Whenever a struct is decorated with the `InlineArray` it is possible to *access* the *array* using a (single dimensional index). Roslyn enforces that when a struct is decorated with the `InlineArray` attribute it is only allowed to contain a single instance field, which is used to *determine* the element type of the array.
Inline arrays doesn't behave fully ordinary arrays. That is, the above inline array can't be used as `int[]` array. There only exist implicit conversion `Span` and `ReadOnlySpan` and we can't do the same kind of type equivalence and replacement.
To that end, we have decided to introduce a new type kind to denote inline arrays as they don't behave like structs nor arrays. However, accessing the array behaves like accessing an ordinary array and we are using the existing functionality here.
With these design decisions, it turns our that
- We can easily identify inline arrays and their element type in QL.
- The dataflow library works out of the box for inline arrays.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/Type.cs:
```diff
@@ -160,6 +168,22 @@ protected void PopulateType(TextWriter trapFile, bool constructUnderlyingTupleTy
+                    .Type;
+
+                var t = Create(Context, elementType);
+                trapFile.inline_array_element_type(this, t);
```
Couldn't we also extract the length of the inline array?

michaelnebel replied:
Yes, that would make sense!

tamasvajk reviewed (commented):
Have you considered implementing `InlineArrayType` with pure QL changes? One could identify relevant structs by looking at their attributes in QL.

How do we represent inline arrays in MaD rows? Can we use `.Element` with inline array types?

michaelnebel commented:
> Have you considered implementing `InlineArrayType` with pure QL changes? One could identify relevant structs by looking at their attributes in QL.
> 
> How do we represent inline arrays in MaD rows? Can we use `.Element` with inline array types?

Yes, I thought about doing that as well, but the nature of an inline array is very different from a struct and it is my impression that the inline array implementation is just a "hack" that is put on structs, so I wanted to keep it separate.
Do you have a strong opinion one way or the other?

I expect that `.Elememt` will work as array read and stores are based on `ArrayAccess` (which we have ensured in the extractor implementation), but I will check. It is my impression that the intention with inline arrays are to mostly do very local performance optimisations and not as such as expose these via an API.

tamasvajk commented:
> Do you have a strong opinion one way or the other?

I don't have a strong opinion. I tend to avoid DB changes if possible, just because it's cumbersome to add the stats, write downgrade scripts, ... But those are already done in this case, so I don't have a clear preference. At the same time, I think the below could get us pretty far without the added DB structure.

```ql
class InlineArrayType extends ValueType {
  Attribute a;
  Field f;
  int length;

  InlineArrayType() {
    a = this.(Attributable).getAnAttribute() and
    a.getType().hasFullyQualifiedName("System.Runtime.CompilerServices", "InlineArrayAttribute") and
    length = a.getArgument(0).getValue().toInt() and
    f = this.getAField()
  }

  Type getElementType() { result = f.getType() }

  ...
}
```

michaelnebel commented:
> ```ql
>     a.getType().hasFullyQualifiedName("System.Runtime.CompilerServices", "InlineArrayAttribute") and
>     length = a.getArgument(0).getValue().toInt() and
>     f = this.getAField()
> ```

This is an excellent point for avoiding to add the extra relation. I will try out something like this, but I would prefer that we introduce the new type kind.

michaelnebel commented:
> > Do you have a strong opinion one way or the other?
> 
> I don't have a strong opinion. I tend to avoid DB changes if possible, just because it's cumbersome to add the stats, write downgrade scripts, ... But those are already done in this case, so I don't have a clear preference. At the same time, I think the below could get us pretty far without the added DB structure.
> 
> ```ql
> class InlineArrayType extends ValueType {
>   Attribute a;
>   Field f;
>   int length;
> 
>   InlineArrayType() {
>     a = this.(Attributable).getAnAttribute() and
>     a.getType().hasFullyQualifiedName("System.Runtime.CompilerServices", "InlineArrayAttribute") and
>     length = a.getArgument(0).getValue().toInt() and
>     f = this.getAField()
>   }
> 
>   Type getElementType() { result = f.getType() }
> 
>   ...
> }
> ```

This is an excellent point for avoiding to add the extra relation. I will try out something like this, but I would prefer that we introduce the new type kind.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs:
```diff
@@ -524,6 +524,13 @@ private static void BuildNamedTypeDisplayName(this INamedTypeSymbol namedType, C
+        public static bool IsInlineArray(this ITypeSymbol type)
+        {
+            var attributes = type.GetAttributes();
+            var isInline = attributes.Any(attribute => attribute.AttributeClass?.Name == "InlineArrayAttribute");
```
Should the namespace be checked too?

michaelnebel replied:
Yes, that would also make sense :-)

tamasvajk commented on csharp/ql/test/library-tests/types/types.cs:
```diff
@@ -56,4 +56,15 @@ interface GenericInterface<T> { }
+    [System.Runtime.CompilerServices.InlineArray(5)]
+    public struct MyMultiDimensionalInlineArray
+    {
+        private MyInlineArray myMultiDimentionalInlineArrayElements;
```
Could you add test cases with `int[][]` and `int[,]` as inline array field types too? That would test the third case of `getDimension()`.

michaelnebel replied:
Sure! I will add some more test coverage! 😄

tamasvajk reviewed (approved):
LGTM

michaelnebel commented:
@tamasvajk: Since you looked last time, the following has changed
- Removed the override of `getChild` on `InlineArrayType` (this was a mistake as the struct can contain more children and as this interferes with the printing of the type (then the type will be printed as a generic)). This was added as a fixup to the QL implementation (and the same goes for the expected test output).
- Added more type related tests.
- Added a models as data test for inline arrays.

When the CI passes, I will start a CI run and if that passes put the PR ready for review.
Thank you for all the feedback you have already provided - this has been really helpful!!!

michaelnebel commented:
DCA looks good (1) No changes to performance (2) There are some minor alert wobliness (which is caused by the different extractor compilations).

tamasvajk reviewed (approved):
LGTM
